### PR TITLE
Python 3

### PR DESCRIPTION
**Pull request by squiddy**
Needs https://github.com/fifengine/fifengine/tree/python3.

untitled
# TODO
- [x] Make tests work again. Our fife import hack fails with metaclass conflicts.
- [ ] Think of a better way to mock fife for unit/integration tests

Fix failing tests
- [ ] tests.game.test_savegame_upgrade:test_upgrade
- [ ] tests.game.test_traderoute:test_traderoute_basic
- [x] tests.game.test_load_save:test_hunter_save_load
- [x] Improve parts of the code, e.g. `list(zip(...))[0]` should be replaceable with `next(zip(...))`
- [x] Fix nose stdout capture when running gui tests (fails with InvalidOperation: fileno)
- [x] `Warning: Failed to open /tmp/tmp8cv7hzdq/yamldata.cache as cache: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte`
- [x] `TypeError: 'float' object cannot be interpreted as an integer` shapes.circle:67. Probably / vs. //
- [ ] Lots of `Object 117269 of type 1 does not know when to pay its rent. Disregard this when loading old savegames or on running cost changes.` messages during tests
- [ ] Check uses of division operator

### CODE REVIEW

**Comment by squiddy** [comment]
As of now:

```
Ran 85 tests in 584.806s
FAILED (SKIP=3, errors=21)

Ran 168 tests in 53.457s
FAILED (SKIP=2, errors=2, failures=2)
```

**Comment by squiddy** [comment]
More progress on the gui tests. Of those 5 errors, 3 are multiplayer related.

```
Ran 85 tests in 820.158s
FAILED (SKIP=3, errors=5)
```

**Comment by squiddy** [comment]
At this point, all gui tests succeed except the ones that fail on python 2 as well.

**Comment by LinuxDonald** [comment]
[linuxdonald@linuxdonald-pc unknown-horizons]$ ./run_uh.py
Recreating all atlases
libGL: pci id for fd 8: 1002:67b0, driver radeonsi
libGL: OpenDriver: trying /usr/lib/xorg/modules/dri/tls/radeonsi_dri.so
libGL: OpenDriver: trying /usr/lib/xorg/modules/dri/radeonsi_dri.so
libGL: Using DRI3 for screen 0
connect(2) call to /dev/shm/jack-1000/default/jack_0 failed (err=No such file or directory)
attempt to connect to server failed
Traceback (most recent call last):
  File "./run_uh.py", line 382, in <module>
    main()
  File "./run_uh.py", line 191, in main
    ret = horizons.main.start(options)
  File "/home/linuxdonald/dev/unknown-horizons/horizons/main.py", line 281, in start
    horizons.globals.fife.run()
  File "/home/linuxdonald/dev/unknown-horizons/horizons/engine/engine.py", line 278, in run
    self.loop()
  File "/home/linuxdonald/dev/unknown-horizons/horizons/engine/engine.py", line 290, in loop
    f()
  File "/home/linuxdonald/dev/unknown-horizons/horizons/timer.py", line 113, in check_tick
    f(self.tick_next_id)
  File "/home/linuxdonald/dev/unknown-horizons/horizons/scheduler.py", line 96, in tick
    callback.callback()
  File "/home/linuxdonald/dev/unknown-horizons/horizons/util/python/callback.py", line 48, in __call__
    return self.callback(*self.args, **self.kwargs)
  File "/home/linuxdonald/dev/unknown-horizons/horizons/ai/aiplayer/__init__.py", line 321, in tick
    self.handle_settlements()
  File "/home/linuxdonald/dev/unknown-horizons/horizons/ai/aiplayer/__init__.py", line 350, in handle_settlements
    result = goal.execute()
  File "/home/linuxdonald/dev/unknown-horizons/horizons/ai/aiplayer/goal/productionchaingoal.py", line 40, in execute
    result = self.chain.build(self._needed_amount)
  File "/home/linuxdonald/dev/unknown-horizons/horizons/ai/aiplayer/productionchain.py", line 95, in build
    return self.chain.build(amount)
  File "/home/linuxdonald/dev/unknown-horizons/horizons/ai/aiplayer/productionchain.py", line 187, in build
    for option in zip(*sorted(expected_costs))[2]:
TypeError: 'zip' object is not subscriptable
AL lib: (EE) alc_cleanup: 1 device not closed

I got this crash. It have something to do with the AI.

**Comment by jorgesumle** [comment]
We shouldn't change the Shebang until we make sure that the game is 100% compatible with Python 3. For the time being, I've never encounter a GNU/Linux distribution that uses Python 3 as the default. According to https://www.python.org/dev/peps/pep-0394/#id13, Arch is the only one.

> however, end users should be aware that python refers to python3 on at least Arch Linux (that change is what prompted the creation of this PEP), so **python should be used in the shebang line only for scripts that are source compatible with both Python 2 and 3**.

**Comment by LinuxDonald** [comment]
Without shebang change uh must always started with: python3 run-uh.py

**Comment by LinuxDonald** [comment]
New ai crash:

``Traceback (most recent call last):
  File "./run_uh.py", line 383, in <module>
    main()
  File "./run_uh.py", line 192, in main
    ret = horizons.main.start(options)
  File "/home/linuxdonald/dev/unknown-horizons/horizons/main.py", line 281, in start
    horizons.globals.fife.run()
  File "/home/linuxdonald/dev/unknown-horizons/horizons/engine/engine.py", line 278, in run
    self.loop()
  File "/home/linuxdonald/dev/unknown-horizons/horizons/engine/engine.py", line 290, in loop
    f()
  File "/home/linuxdonald/dev/unknown-horizons/horizons/timer.py", line 113, in check_tick
    f(self.tick_next_id)
  File "/home/linuxdonald/dev/unknown-horizons/horizons/scheduler.py", line 96, in tick
    callback.callback()
  File "/home/linuxdonald/dev/unknown-horizons/horizons/util/python/callback.py", line 48, in __call__
    return self.callback(*self.args, **self.kwargs)
  File "/home/linuxdonald/dev/unknown-horizons/horizons/ai/aiplayer/__init__.py", line 323, in tick
    self.international_trade_manager.tick()
  File "/home/linuxdonald/dev/unknown-horizons/horizons/ai/aiplayer/internationaltrademanager.py", line 132, in tick
    self._add_route()
  File "/home/linuxdonald/dev/unknown-horizons/horizons/ai/aiplayer/internationaltrademanager.py", line 128, in _add_route
    bought_resource, sold_resource, settlement, settlement_manager = max(final_options)[1:]
TypeError: unorderable types: int() > NoneType()
AL lib: (EE) alc_cleanup: 1 device not closed
``

**Comment by LinuxDonald** [comment]
fixed both crashs Big THX goes to @MasterofJOKers for the fixes :)

File: horizons/ai/aiplayer/internationaltrademanager.py [hunk 1]
```diff
@@ -125,7 +125,7 @@ def _add_route(self):
 			total_value = (best_sale[0] if best_sale else 0) + (best_buy[1] if best_buy else 0) * buy_coefficient
 			final_options.append((total_value, best_buy[2] if best_buy else None, best_sale[2] if best_sale else None, settlement, settlement_manager))
 
-		bought_resource, sold_resource, settlement, settlement_manager = max(final_options)[1:]
+		bought_resource, sold_resource, settlement, settlement_manager = sorted(final_options, reverse=True, key=lambda x: x[0])[0][1:]
```
Comment (LinuxDonald) [inline horizons/ai/aiplayer/internationaltrademanager.py]: @squidds this change is from MasterofJOKers I only added it to git.

**Comment by LinuxDonald** [comment]
@squiddy the tearing (destroy) is broken on py3 branch. It dont destroy buildings. When i go over tress they get transparent. But on buildings not. When i click nothing happen on all type of building/tress usw.

**Comment by squiddy** [comment]
@LinuxDonald I can destroy buildings & trees just fine. There might be an issue with displaying them transparent. We do have a GUI test for the tearing tool, however it doesn't check if something was actually removed. I'm going to extend the test.

**Comment by LinuxDonald** [comment]
tested it with an new savegame now it works again.

**Comment by squiddy** [comment]
@LinuxDonald I'm pretty sure loading savegames created with python2 is asking for trouble, especially older ones. For example before savegame version 57 we were using pickle to serialize scenario_variables. I don't think pickled data from python2 can be read by python3 out-of-the-box.

This also affects other uses of pickle, such as: yamlcache and atlas. Although removing both and have them be recreated works fine.

**Comment by squiddy** [comment]
Current test status: all succeeding gui tests in python2 work in python3 as well. Long AI tests too.

Failing game tests:

tests.game.test_savegame_upgrade:test_upgrade
tests.game.test_traderoute:test_traderoute_basic
tests.game.test_load_save:test_hunter_save_load

The first two are using a savegame, so that might be an issue.

**Comment by squiddy** [comment]
If we still want to go through with this, I'll try to get these changes on current master sometime next week. IIRC there are only minor issues left, so I hope we can get this into master soon.

**Comment by LinuxDonald** [comment]
@squiddy i heard today that fife python3 branch have some problems with pychan. But beliar want to look at it.

**Comment by squiddy** [comment]
I see. Perhaps we should wait with a merge until the python3 port in fife is ready.

**Comment by LinuxDonald** [comment]
Yes. But an updatet py3 branch of uh would be cool for testing and find bugs on fife py3 branch :)

**Comment by squiddy** [comment]
Is there a plan when Python 3 support for FIFE will land in master? Because I don't feel like rebasing this branch every couple months.

**Comment by MasterofJOKers** [comment]
While fife's py3 branch worked for py3, it did not work for py2 any more. There is currently work going on to re-convert fife's py2-code to support py2 and py3 (using futurize instead of 2to3 directly). So I'd say, if you rebase onto master next week, there's probably py3 support in fife by that time.

**Comment by LinuxDonald** [comment]
The current plan is port UH to fife pychan_port and then pychan_port will get merged into master. After this python3 will get tested and merged too. We need first port it to pychan_port because testing fife if there are some bugs.

https://github.com/unknown-horizons/unknown-horizons/pull/2601

**Comment by squiddy** [comment]
If that port needs many changes in

**Comment by LinuxDonald** [comment]
@helios2000 is working on new render. And fife team plan to switch Editor to Tiled Map Editor.
does i forgot something? @helios2000 

@squiddy i would need your help at pychan port on UH side because there are needed python changes too.

Current Status of the gui:

https://abload.de/img/2017-01-22.00-58-58.2rest0.png

https://abload.de/img/2017-01-22.00-58-48.4wzswo.png

**Comment by squiddy** [comment]
@LinuxDonald Sorry, I doubt I have time. In the future it might be better to announce future plans a bit earlier. I'm don't even know what this pychan branch is about, only that now I'm going to wait until it's done before I continue working on this to avoid possible merge conflicts.

**Comment by LinuxDonald** [comment]
@squiddy you can start port our pychan_port branch to python3 :-)

**Comment by squiddy** [comment]
Thanks. But I'd rather start once it is in master.

**Comment by AndyMender** [comment]
I can definitely join in on this one as I'm more familiar with Python3 and there is quite some nice tricks would could do while rebasing :).

**Comment by LinuxDonald** [comment]
Python3 port can now be started. Pychan changes are merged.

**Comment by squiddy** [comment]
> Python3 port can now be started. Pychan changes are merged.

Just to make sure, does that mean that Python3 support in fife is merged as well?

@AndyMender 

Let's join forces on this one. The approach I took here is outlined in the first couple commits, it's mostly running `2to3` and fixing everything that breaks afterwards. I think we should restart that work based on the current UH master (I'd rather not rebase this branch).

Most changes are pretty straight forward, there is one tricky part though:

In order to run unit/game tests, we've completely mocked away the fife module, otherwise we would need to setup the engine etc., which would slow them down significantly. By using the `Dummy` object in `horizons.ext.dummy` this worked pretty good in python2, but not so in python3. I've had to resort to things like this (https://github.com/unknown-horizons/unknown-horizons/pull/2490/commits/a8c03613eacc9e4685c439da0626624c82d7ad44#diff-ca93f5c2b15c9af9def6ea476ae494c7) in order to make it work.

I think that's the biggest hurdle to this port, 2to3 does most of the work, and making existing code more idiomatic / cleanup is more diligence work, and can probably done once this is in master.

**Comment by Beliaar** [comment]
@squiddy Nope, what @LinuxDonald means is that I updated the python3 branch to fife master, which included the pychan changes.

**Comment by AndyMender** [comment]
If I understand correctly, we first close/delete the existing Python3 branch and create an entirely new one from the current master branch, yes?

I agree on using 2to3, though I think it's also a good chance for us to trim off some of the redundancies in the code, like superfluous variable assignments, etc. We might potentially have issues with old-style vs new-style class mechanics, but we'll deal with it when the time comes :).

Would you like to open a new branch then? I am ready whenever :).

**Comment by squiddy** [comment]
> @squiddy Nope, what @LinuxDonald means is that I updated the python3 branch to fife master, which included the pychan changes.

That's unfortunate. @LinuxDonald perhaps I'm in the wrong here, but I want to get the port to python3 into master as soon as possible, and not have it as a long-standing branch again. The alternative is that UH master will only work with the fife python3 branch.

@AndyMender 

> though I think it's also a good chance for us to trim off some of the redundancies in the code, like superfluous variable assignments, etc. We might potentially have issues with old-style vs new-style class mechanics, but we'll deal with it when the time comes :).

I'd like to get the port *working* as fast as possible, the time for cleanup can come later IMO. Otherwise we risk spending too much time on code cleanup rather than getting the port itself complete.

**Comment by squiddy** [comment]
@AndyMender Let me word that differently. The way I see it, priority #1 is to get the majority of tests working. This will then help a lot when refactoring/cleanup code.

I'm going to push a new branch (with the changes from here, a bit cleaned up) later. We can then split work properly. :)

**Comment by Beliaar** [comment]
@squiddy  @LinuxDonald Well, I am not sure how others think, but I would not merge the branch into master before the changes are tested by several people. It works on my end, but that does not mean that there is not someone else that may do things that may not work.

**Comment by squiddy** [comment]
@Beliaar I may have worded that ambigiously, but I was talking about the python3 branch of UH. If FIFE doesn't yet want to merge the python3 port yet, that's totally fine by me and I understand that.

However, I don't want our efforts in porting UH to linger around needlessly and potentially force us to rebase continously. If UH master were to only work with python3 starting from now, we'd have the opportunity to get that into the hands of players and provide needed feedback to FIFE.

What do you think @LinuxDonald ?

**Comment by Beliaar** [comment]
@squiddy Ah, of course, if it is no problem for you to work on a branch of fife instead of master. This should, of course, speed up the testing process, though.

**Comment by squiddy** [comment]
I've tried to split up the tasks that need to be done and put them in a separate project: https://github.com/unknown-horizons/unknown-horizons/projects/2

**Comment by LinuxDonald** [comment]
When UH python3 is done it will get merged into master. We need UH python3 version to make the python3 version of fife final. Helios2k want to see python3 support merged on fife side.

**Comment by LinuxDonald** [comment]
@squiddy should i add @AndyMender to UH Team that you both can work on the same branch? or do you want prs from him?

**Comment by squiddy** [comment]
I'd prefer pull requests.

**Comment by AndyMender** [comment]
I'd also prefer pull requests at least until both parties are sure I can continuously contribute to the project :). I have halted my C programming related activities to focus fully on the UH project.

**Comment by LinuxDonald** [comment]
Closing because new python 3 branch will get created

**Pull request closed without merging**